### PR TITLE
Enable --nv option to bind in nvidia drivers

### LINKED ISSUE #1886: Need to implement --nv option
## Version of Singularity:

```
$ singularity --version
singularity version 3.0.0-alpha.1-117-gc297513f
```

### Expected behavior

The `--nv` option will expose the GPU like so:
```
$ singularity exec --nv my_container some command
```

### Actual behavior

The `--nv` option doesn't really do anything yet.

### Steps to reproduce behavior

Try using the `--nv` option on a host system with a GPU.

### PR DESCRIPTION
**Pull request by ctmadison**
**Description of the Pull Request (PR):**

Add in support for automatically binding Nvidia libraries/drivers with a cmd line --nv argument.

**This fixes or addresses the following GitHub issues:**

- Fixes #1886

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [ ] I have tested this PR locally with a `make test`
- [x] This PR is against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge

Attn: @singularityware-admin

### CODE REVIEW

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]:
```
if err != nil {
    return nil, fmt.Errorf("error running nvidia-container-cli: %v", err)
}
```

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]:
```
if err != nil {
    return nil, fmt.Errorf("nvidia-container-cli not found: %v", err)
}
```

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]: `func GetNvidiaBindPath(abspath string) ([]string, error) {`

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]:
```
if err != nil {
    return nil, fmt.Errorf("filepath.Glob() failed: %v", err)
}
```

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]:
```
if err != nil {
    return nil, fmt.Errorf("failed to open file %v: %v", filename, err)
}
```

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]:
```
if err != nil {
    return nil, fmt.Errorf("ldconfig not in path: %v", err)
}
```

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]:
```
if err != nil {
    return nil, fmt.Errorf("failed running ldconfig -p: %v", err)
}
```

File: src/cmd/singularity/cli/actions.go
Comment (bauerm97) [inline src/cmd/singularity/cli/actions.go]:
```
NvidiaBindPaths, err := nvidiautils.GetNvidiaBindPath(buildcfg.SINGULARITY_CONFDIR)
if err != nil {
    sylog.Fatalf("Unable to capture nvidia bind points: %v", err)
}
```

**Review by bauerm97 — CHANGES_REQUESTED** [review]
This needs some readability tweaks before I'll be able to dive in further, once you fix what I've pointed out I'll take a second glance

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go]: I'm not expert on regular expressions, but this seems like it is saying "if there is a `#` anywhere in the line it's a comment".  

Shouldn't it be like "`^#`" or something?

File: src/cmd/singularity/cli/actions.go
Comment (GodloveD) [reply src/cmd/singularity/cli/actions.go]: Maybe it should be a warning instead of a failure though?  That is the way 2.x works.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: That's what I was thinking at first too @bauerm97 but I don't think that will work in this instance.  If singularity cannot find `nvidia-container-cli` or if `nvidia-container-cli` returns an error it should just issue a warning and then move on to try to use the conf file instead.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: Same comment as above.  As much as I dislike it, I think all of this code actually does need to be wrapped in the if statement b/c we don't want to bail if it doesn't work.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: I agree with this though I think it should just spit a warning and then the function should return.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: I agree with this though I think it should just spit a warning and then the function should return.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: I agree with this though I think it should just spit a warning and then the function should return.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: I agree with this though I think it should just spit a warning and then the function should return.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go]: appending to a string array in a loop sounds like it will be very slow to me.  I think there are faster ways to do this?  But I don't want us spending time on premature optimization if this is already pretty quick.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go]: Is it necessary to have the second variable `strArray` here?  Can `cliEntries` be constructed from `bindArray`?

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: Hmmm... nvidia-container-cli is an optional component. Don't want to return if error.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: Hmmm... nvidia-container-cli is an optional component. Don't want to return if error.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go]: Oh so we are now using the _union_ of what we find in the conf file and what `nvidia-container-cli` yields?  I'm not sure I like that.  Kindof think we should default to `nvidia-container-cli` output and only fall back to conf file if it fails.

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]
`make install` should cp `etc/nvliblist.conf`  to `$PREFIX/etc/nvliblist.conf`

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [reply src/pkg/util/nvidiautils/nvidiautils.go]: Definitely want to error if we're unable to execute `nvidia-container-cli` if it does exist in `$PATH`

**Review by GodloveD — CHANGES_REQUESTED** [review]
I'm also not done looking at this yet, but in the interest of time I'm making a first round of comments.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [reply src/pkg/util/nvidiautils/nvidiautils.go]: Fair enough.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: No I disagree. `nvidia-container-cli` is still in a pretty early state afaik.  They have already changed he cli on me once.  If the user has a newer version of the binary installed that doesn't do what we expect we still need the option of falling back to the conf file instead.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [reply src/pkg/util/nvidiautils/nvidiautils.go]: If you think it should spit a warning, then the person who calls this function can look at the error and determine if it should warn or error

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [reply src/pkg/util/nvidiautils/nvidiautils.go]: In this case, I would actually say to break these into functions:

```
func nvidiaContainerCli() ([]string, error) {
    // use nvidia-container-cli (if present)
    command, err := exec.LookPath("nvidia-container-cli")
    if err =! nil {
        return nil, nil
    }
}      
```

Then within the main code can simply do this:

```
s, err := nvidiaContainerCli()
if err != nil {
    return nil, fmt.Println("Unable to do something...: %v", err)
}
...
```

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go]: why is this in a for loop?  It should only be a single file right?

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go]: I think you can use `defer` here to close when done?  Is that right?

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go]: For a possible simpler and more readable solution check this out (though I don't know if performance would suffer greatly).
https://golang.org/pkg/strings/#HasPrefix

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: Ok, I like this better...

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (jmstover) [reply src/pkg/util/nvidiautils/nvidiautils.go]: nvidia-container-cli didn't provide everything though right? I thought there was some stuff missing.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: I don't think so. Because the bind array has entries like /dir/dir/file:/.singularity.d/file in it. I don't think it would be any faster and it would be more work. Are you thinking of a specific method?

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: I don't know another faster way (maybe other than building everything from scratch using buffers, etc.)

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go]: This should already be handled above right?  nvidiaConfFileline should not include empty entries afaik.

https://github.com/singularityware/singularity/blob/672f0e753500b55b26a04f736fb72b3c47ba3216/src/pkg/util/nvidiautils/nvidiautils.go#L36

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: probably....

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: Well, the nvidia-container-cli function only returns the /dir/file.so.xxx.xx entries (where the xxx.xx is the driver version). I can't just strip off the .xxx.xx and use the so entries because some things (like nvidia-smi binary look for the file.so.1 entry). I could make it work if that is the way you think I should go...

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: agreed

**Review by GodloveD — CHANGES_REQUESTED** [review]

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: I think what you want to do is use _either_ the output of `nvidia-container-cli` _or_ the output of the `nvidia.conf` file as a set of things to search the `ld.so.cache` for (through `ldconfig` output).  

Does that make sense?  In other words if you get usable output from `nvidia-container-cli` skip parsing the `nvidia.conf` file and just proceed directly to pattern matching between `nvidia-container-cli` output and `ldconfig` output.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: yes, you are correct.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: In my research, it was recommended to NOT use defer. I had it that way at first but changed it. But I will defer to others with more go experience...

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: If you want to just append `.1` to all of the `nvidia-container-cli` entries and search for them as well, I think it's reasonable?  

@jmstover yes there was a recent issue where a user was confused b/c `nvidia-container-cli` does not provide open GL libraries, but that behavior is correct and our behavior is wrong.  We should comment those libs out of the conf file imho because those libs are not strictly tied to the version of the driver as the other libs are.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: agreed.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (jmstover) [reply src/pkg/util/nvidiautils/nvidiautils.go]: @GodloveD 
Maybe not on _recent_ drivers with libglvnd. But my `375.39` driver version, the GL libs *are* tied to the driver version.

i.e.  libGLX_nvidia.so.375.39

And we had cards where we couldn't update past a specific driver version because they removed support. So... no recent driver to have any of the new fancy stuff.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: Yes, I started using more string. functions after I wrote this code and didn't take time to go back and change it.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [reply src/pkg/util/nvidiautils/nvidiautils.go]: @ctmadison I just had a brain wave.  I think you need to strip driver version from the `nvidia-container-cli` output and then use that as a _prefix_ when searching the output of `ldconfig`.  Tht way you will get any stray `.so.1` files (and other file appends) too.

Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: Yes, that should work. But, there is still the issue that @jmstover brought up about the GL libs that we have to agree on.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]: If we're unable to execute the `nvidia-container-cli` command, we *should* return an error:

```
if err != nil {
    return nil, nil, fmt.Errof("Unable to execute ldconfig: %v", err)
}
```

We should only be ignoring errors in the case where `nvidia-container-cli` doesn't exist in path

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]: If we're only returning errors which *shouldn't* be ignored, then we should `sylog.Fataf(...)` instead.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (bauerm97) [inline src/pkg/util/nvidiautils/nvidiautils.go]: `defer file.Close()`

Comment (ctmadison) [reply src/pkg/util/nvidiautils/nvidiautils.go]: ok

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go:38]: I'm not sure this will work when it gets input that does not include `/` or does include whitespace or whatever, but I cannot test b/c I don't have access to a GPU right now.

File: src/pkg/util/nvidiautils/nvidiautils.go
Comment (GodloveD) [inline src/pkg/util/nvidiautils/nvidiautils.go]: I think the addition of `strings.HasPrefix()` will pick up any `*.so.1` libs or others with different suffixes.  But needs to be tested to be sure.

**Review by GodloveD — DISMISSED** [review]
I looked this over and added a few changes.  If they all work and are acceptable I approve.

**Comment by jmstover** [comment]
`$(SRC)/etc/nvliblist.conf` is not being installed to $(PREFIX)/etc/singularity/nvliblist.conf on `make install`

File: src/cmd/singularity/cli/actions.go
Comment (jmstover) [inline src/cmd/singularity/cli/actions.go]: This should be:  `%v/nvliblist.conf` ...   buldcfg.SINGULARITY_CONFDIR includes the singularity/ directory

File: src/cmd/singularity/cli/actions.go
Comment (jmstover) [inline src/cmd/singularity/cli/actions.go]: Should this be if `len(NvidiaBindPaths)` ?

**Review by jmstover — DISMISSED** [review]

**Comment by ArangoGutierrez** [comment]
Travis is not happy due
```
 CHECK go lint
/home/travis/gopath/src/github.com/sylabs/singularity/src/pkg/util/nvidiautils/nvidiautils.go:27:31: error strings should not be capitalized or end with punctuation or a newline
/home/travis/gopath/src/github.com/sylabs/singularity/src/pkg/util/nvidiautils/nvidiautils.go:34:31: error strings should not be capitalized or end with punctuation or a newline
/home/travis/gopath/src/github.com/sylabs/singularity/src/pkg/util/nvidiautils/nvidiautils.go:51:31: error strings should not be capitalized or end with punctuation or a newline
```

**Review by jmstover — APPROVED** [review]
This is fully working for me now.

**Review by ArangoGutierrez — APPROVED** [review]

**Comment by ArangoGutierrez** [comment]
this img was rendered using the `--nv`

**Comment by ArangoGutierrez** [comment]
```bash
[eduardo@rhel7 tmp]$ nvidia-container-[TAB]
nvidia-container-cli           nvidia-container-runtime       nvidia-container-runtime-hook  
[eduardo@rhel7 tmp]$ singularity  exec --nv docker://ubuntu nvidia-smi
Fri Sep 21 21:13:47 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 396.37                 Driver Version: 396.37                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Quadro M620         Off  | 00000000:01:00.0  On |                  N/A |
| N/A   51C    P8    N/A /  N/A |   1055MiB /  1999MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      2099      G   /usr/bin/X                                   229MiB |
|    0      2840      G   /usr/bin/gnome-shell                         354MiB |
|    0      3598      G   ...-token=06CB68BCB817E1EXXXXXXXXXXXXXXXXX    90MiB |
|    0      3788      G   ...uest-channel-token=17275785424XXXXXXXXX   134MiB |
|    0      4275      G   ...-token=5848D7EB3EABAXXXXXXXXXXXXXXXXXXX    74MiB |
|    0     30710      G   ...-type=gpu-process --disable-features=Co   166MiB |
+-----------------------------------------------------------------------------+
```

**Review by GodloveD — APPROVED** [review]
This looks good to me now.

**Review by bauerm97 — APPROVED** [review]

**Comment by bauerm97** [comment]
@GodloveD Good to merge once Travis passes

**Comment by GodloveD** [comment]
Thanks @ctmadison @jmstover @ArangoGutierrez and @bauerm97 !

**Pull request merged**
